oak_channel/SPEC.md

```diff
@@ -43,244 +42,157 @@ The service MUST NOT send request messages.
-
-Messages are parsed from frames by appending the frame bodies in the order they
-are received.
-
-## Examples for Illustration
-
-### Frame Flags
-
-For messages sent as more than two frames, the first frame would have the start
-flag set, all other flags would be unset. The last frame would have the end flag
-set, all other flags would be unset. Any frames in between would have no flags
-set.
-
-For small messages sent as a single frame, the start flag and the end flag would
-be set, all other flags would be unset.
-
-### Layers Diagram
-
-The diagram below shows a practial example of how the request half of an
-invocation would look like after being encoded as two frames.
-
-#### Frame 0
-
-```text
- 0                   1                   2                   3
- 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                         frame padding                         |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|          frame length         |          frame flags          |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|             frame body [ request_message length ]             |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|          frame body [ request_message invocation_id ]         |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|            frame body [ request_message method_id ]           |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|             frame body [ request_message padding ]            |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                                                               |
-+                                                               +
-|                                                               |
-+   frame body [ request_message body [ method parameters ] ]   +
-|                                                               |
-+                                                               +
-|                                                               |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ...
-```
+To send a message, the sender MUST first encode the message into frames:
 
-<!-- Diagram generated with https://www.luismg.com/protocol/, using the schema
-"frame padding:32,frame length:16,frame flags:16,frame body [ request_message length ]:32,frame body [ request_message invocation_id ]:32,frame body [ request_message method_id ]:32,frame body [ request_message padding ]:32,frame body [ request_message body [ method parameters ] ]:128" -->
+The sender MUST chunk byte representation of the message into a set of frame
+bodies. The last frame body of the resulting set MAY be shorter than 4,080 bytes
+(the maximum length of a frame body). Any preceding frame bodies MUST be exactly
+4,080 bytes in length. The sender MUST then create a frame from each frame body.
```

Comment:
I would like to leave it to the client to decide what framing size to use, and just set the upper bound. And we should handle frames of different sizes within the same message too IMO.

Reply:
If full frames don't have a fixed size the sender will need to put the number of frames and the message size in the header if we want to be able to reserve the appropriate amount or memory and know when we have received all the frames. I will add an additional field to the header to support that.

Reply:
Isn't the message size (which is already there) enough?

Reply:
It is enough to reserve the space, but since we are not guaranteed that each full frame will have the same size we cannot calculate the expected number of frames from the message size.

Reply:
But I thought in practice the logic on the receiving side is to allocate the space, then append each frame in order until it's full. We don't need to know the number of frames ahead of time, just check that the total size adds up correctly.

Reply:
Let's step back for a moment. Is the issue that we risk missing a frame within a message and not notice? Or that individual bits may be corrupted? If the former, then a flag indicating the last frame of a message may be enough. For the latter, I agree some sort of checksum may be useful, though that may be also done at the application layer in principle.

Reply:
The important part is to have very high confidence that we will detect frame misalignment or corruption of the header, as this will indicate an unrecoverable situation for the channel and we should terminate the workload. Knowing an exact expected value (or small valid range) for each header field for a specific frame helps a lot with this.

If any bits in the body are corrupted we will detect it (at least for encrypted user requests) when we try to decrypt it, but that should only affect the single message that is corruped, not the stream as a whole, so we won't have to terminate the workload.

In my previous message I mentioned a checksum for the message, but actually meant a checksum for the frame.

Reply:
Makes sense. So perhaps have a small checksum and also flags for each header? similar to TCP.

Reply:
+1, I'd like to see some kind of checksum. Memory corruption, even inside one machine, is all too common.

Reply:
I don't think out-of-order frames or entire frames going missing is likely to happen due to data corruption, so we probably don't need frame numbers if we have a checksum to validate the header of the frame and rely on message size to know when we have received everything.

If we leave message integrity up to the application (e.g. since we are using authenticated encryption for user requests) we only really need the checksum to cover the header:

- Version
- Frame size
- Message Size
- Invocation ID
- Header Checksum

The header checksum could be something simple like the result or xoring previous batches of 4 bytes from the header, or it could be a truncated hash.

WDYT?